### PR TITLE
Support Anonymous cart in customer

### PR DESCRIPTION
**Pull request by barbara79**
# Description

Closes #

# Checklist

- [x] Update release Notes
- [ ] Add to the current github milestone 
- [x] Update commercetools api reference

### CODE REVIEW

File: commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerInStoreSignInCommand.java [hunk 1]
```diff
@@ -28,11 +37,12 @@
 
     private final String storeKey;
 
-    private CustomerInStoreSignInCommand(final String storeKey, final String email, final String password, @Nullable final String anonymousCartId, @Nullable final String anonymousId, final AnonymousCartSignInMode anonymousCartSignInMode,
+    private CustomerInStoreSignInCommand(final String storeKey, final String email, final String password, @Nullable ResourceIdentifier<Cart> anonymousCart, @Nullable final String anonymousCartId, @Nullable final String anonymousId, final AnonymousCartSignInMode anonymousCartSignInMode,
```
Comment (jenschude) [inline commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerInStoreSignInCommand.java:40]:
```suggestion
    private CustomerInStoreSignInCommand(final String storeKey, final String email, final String password, @Nullable ResourceIdentifier<Cart> anonymousCart, @Nullable final String anonymousId, final AnonymousCartSignInMode anonymousCartSignInMode,
```

As this is a private constructor you can change the signature and completely drop the anonymousCartId. Now you have to refactor the other methods to create a ResourceIdentifier if needed.

File: commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerSignInCommand.java [hunk 2]
```diff
@@ -37,20 +39,27 @@
...
+    private final ResourceIdentifier<Cart> anonymousCart;
+    @Nullable
     private final String anonymousId;
     @Nullable
     private final AnonymousCartSignInMode anonymousCartSignInMode;
     @Nullable
     private final Boolean updateProductData;
 
-    private CustomerSignInCommand(final String email, final String password, @Nullable final String anonymousCartId, @Nullable final String anonymousId, final AnonymousCartSignInMode anonymousCartSignInMode,
+    private CustomerSignInCommand(final String email, final String password, @Nullable final String anonymousCartId, @Nullable ResourceIdentifier<Cart> anonymousCart, @Nullable final String anonymousId, final AnonymousCartSignInMode anonymousCartSignInMode,
```
Comment (jenschude) [inline commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerSignInCommand.java:57]: same here. Drop the anonymousCartId and refactor the other methods to use create a ResourceIdentifier if needed

```suggestion
    private CustomerSignInCommand(final String email, final String password, @Nullable ResourceIdentifier<Cart> anonymousCart, @Nullable final String anonymousId, final AnonymousCartSignInMode anonymousCartSignInMode,
```

**Review by jenschude — CHANGES_REQUESTED** [review]

File: commercetools-models/src/main/java/io/sphere/sdk/customers/CustomerDraftDsl.java [hunk 3]
```diff
@@ -29,11 +30,17 @@ public static CustomerDraftDsl of(final CustomerName customerName, final String
...
+    @Deprecated
     public CustomerDraftDsl withCart(final io.sphere.sdk.carts.Cart cart) {
         Objects.requireNonNull(cart);
         return withAnonymousCartId(cart.getId());
     }
 
+    public CustomerDraftDsl withCartAnonymousCart(final io.sphere.sdk.carts.Cart cart) {
+        Objects.requireNonNull(cart);
+        return withAnonymousCart(cart.toResourceIdentifier());
+    }
```
[the remark this answers is not in the document]
Comment (barbara79) [reply commercetools-models/src/main/java/io/sphere/sdk/customers/CustomerDraftDsl.java:42]: sure! I didn't think about the Referenceable<Cart>  ;)

File: commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerInStoreSignInCommand.java (hunk 1 above)
Comment (barbara79) [reply commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerInStoreSignInCommand.java:40]: ok thanks I'm gonna change it

**Review by jenschude — APPROVED** [review]

File: commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerSignInCommand.java [hunk 4]
```diff
@@ -114,7 +116,7 @@ public String getPassword() {
     @Deprecated
     @Nullable
     public String getAnonymousCartId() {
-        return anonymousCart.getId();
+        return Optional.ofNullable(anonymousCart).map(ResourceIdentifier::getId).orElse(null);
```
Comment (barbara79) [inline commercetools-models/src/main/java/io/sphere/sdk/customers/commands/CustomerSignInCommand.java:119]: 👍

**Review by jenschude — APPROVED** [review]

**Pull request merged**